homeassistant/components/energyzero/sensor.py

```diff
@@ -0,0 +1,215 @@
+        name="Current percentage of highest electricity price - Today",
+        native_unit_of_measurement=PERCENTAGE,
+        icon="mdi:percent",
+        value_fn=lambda data: data[SERVICE_ENERGY_TODAY].pct_of_max_price,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up EnergyZero Sensors based on a config entry."""
+    coordinator: EnergyZeroDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    entities: list[EnergyZeroSensorEntity] = []
+    entities.extend(
+        EnergyZeroSensorEntity(
+            coordinator=coordinator,
+            description=description,
+            name="Energy Market Price",
+            service="today_energy",
+        )
+        for description in SENSORS_ENERGY
+    )
+    if entry.data[CONF_GAS]:
+        entities.extend(
+            EnergyZeroSensorEntity(
+                coordinator=coordinator,
+                description=description,
+                name="Gas Market Price",
+                service="today_gas",
+            )
+            for description in SENSORS_GAS
+        )
+    async_add_entities(entities)
+
+
+class EnergyZeroSensorEntity(
+    CoordinatorEntity[EnergyZeroDataUpdateCoordinator], SensorEntity
+):
+    """Defines a EnergyZero sensor."""
+
+    _attr_attribution = ATTRIBUTION
+    entity_description: EnergyZeroSensorEntityDescription
+
+    def __init__(
+        self,
+        *,
+        coordinator: EnergyZeroDataUpdateCoordinator,
+        description: EnergyZeroSensorEntityDescription,
+        name: str,
+        service: str,
+    ) -> None:
+        """Initialize EnergyZero sensor."""
+        super().__init__(coordinator=coordinator)
+        self.entity_id = f"{SENSOR_DOMAIN}.energyzero_{service}_{description.key}"
+        self.entity_description = description
+        self._attr_unique_id = (
+            f"{coordinator.config_entry.entry_id}_{service}_{description.key}"
+        )
+
+        self._attr_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, f"{coordinator.config_entry.entry_id}_{service}")},
+            manufacturer="EnergyZero",
+            name=name,
+        )
+
+    @property
+    def native_value(self) -> float | datetime | None:
+        """Return the state of the sensor."""
+        return self.entity_description.value_fn(self.coordinator.data)
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return the state attributes."""
+        if self.entity_description.key == "average_price":
+            return {
+                "today": self.coordinator.data[SERVICE_ENERGY_TODAY].timestamp_prices,
+                "tomorrow": self.coordinator.data[
+                    SERVICE_ENERGY_TOMORROW
+                ].timestamp_prices
+                if self.coordinator.data[SERVICE_ENERGY_TOMORROW]
+                else None,
```

Comment:
I approached it this way because I actually want to prevent an entity from giving None every time between 00:00 and 14:00.
Inspired by how it was done with this custom integration: https://github.com/JaccoR/hass-entso-e

Reply:
Right, but this is not what attributes are for.